src/prng.rs

```diff
@@ -35,48 +37,83 @@ fn random_field_and_seed<F: FieldElement>(length: usize) -> (Vec<F>, Vec<u8>) {
     (data, seed)
 }
 
-fn random_field_from_seed<F: FieldElement>(seed: &[u8], length: usize) -> Vec<F> {
-    let key = GenericArray::from_slice(&seed[..BLOCK_SIZE]);
-    let nonce = GenericArray::from_slice(&seed[BLOCK_SIZE..]);
-    let mut cipher = Aes128Ctr::new(&key, &nonce);
-
-    let mut output = vec![F::zero(); length];
-    let mut output_written = 0;
-
-    let length_in_blocks = length * F::BYTES / BLOCK_SIZE;
-    // add one more block to account for rejection and roundoff errors
-    let buffer_len_in_blocks = std::cmp::min(MAXIMUM_BUFFER_SIZE_IN_BLOCKS, length_in_blocks + 1);
-    // Note: buffer_len must be a multiple of BLOCK_SIZE, so that different buffer
-    // lengths return the same data
-    let buffer_len = buffer_len_in_blocks * BLOCK_SIZE;
-    let mut buffer = vec![0; buffer_len];
-
-    while output_written < length {
-        // zero the buffer
-        for b in &mut buffer {
-            *b = 0;
-        }
+/// This type implements an iterator that generates a pseudorandom sequence of field elements. The
+/// sequence is derived from the key stream of AES-128 in CTR mode with a random IV.
+pub struct Prng<F: FieldElement> {
+    phantom: PhantomData<F>,
+    cipher: Aes128Ctr,
+    length: usize,
+    buffer: Vec<u8>,
+    buffer_index: usize,
+    output_written: usize,
+}
 
+impl<F: FieldElement> Prng<F> {
+    /// Constructs an iterator over a pseudorandom sequence of field elements of length `length`.
+    /// `seed` is used to seed the underlying pseudorandom number generator.
+    pub fn new_with_length(seed: &[u8], length: usize) -> Self {
+        let key = GenericArray::from_slice(&seed[..BLOCK_SIZE]);
+        let nonce = GenericArray::from_slice(&seed[BLOCK_SIZE..]);
```

Comment:
Done. I added a new error type, PrngError, which is propagated from here. The only caller is `random_field_from_seed`, which is private to this package. It just unwraps the error instead of propagating it, which is consistent with the current behavior, which would panic if the seed were too short.